File: pkg/store/conversation_store_test.go
package store

import (
	"context"
	"errors"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestConversationStore_Create_Success(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "Test Conv")
	err := db.Conversations.Create(ctx, conv)
	require.NoError(t, err)

	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, conv.ID, got.ID)
	assert.Equal(t, conv.Title, got.Title)
	assert.Equal(t, conv.UserID1, got.UserID1)
	assert.Equal(t, conv.UserID2, got.UserID2)
}

func TestConversationStore_Create_DuplicateKey(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv1 := newTestConv(uid(), "userA", "userB", "direct", "Conv 1")
	require.NoError(t, db.Conversations.Create(ctx, conv1))

	// Same unique (user_id1, user_id2) pair → unique constraint violation.
	conv2 := newTestConv(uid(), "userA", "userB", "direct", "Conv 2")
	err := db.Conversations.Create(ctx, conv2)
	assert.ErrorIs(t, err, ErrDuplicateKey)
}

func TestConversationStore_Get_Found(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "Hello")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, conv.ID, got.ID)
	assert.Equal(t, "Hello", got.Title)
}

func TestConversationStore_Get_NotFound(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	_, err := db.Conversations.Get(ctx, "nonexistent-id")
	assert.ErrorIs(t, err, ErrNotFound)
}

func TestConversationStore_GetByUsers_BothOrderings(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), "alice", "bob", "direct", "Alice-Bob")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	// Query with (alice, bob).
	got, err := db.Conversations.GetByUsers(ctx, "alice", "bob")
	require.NoError(t, err)
	assert.Equal(t, conv.ID, got.ID)

	// Query with (bob, alice) — reversed order.
	got, err = db.Conversations.GetByUsers(ctx, "bob", "alice")
	require.NoError(t, err)
	assert.Equal(t, conv.ID, got.ID)
}

func TestConversationStore_GetByUser_OrderAndPagination(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	userA := uid()
	userB := uid()

	now := time.Now()
	// Create 3 conversations with different LastMessageAt.
	conv1 := newTestConv(uid(), userA, userB, "direct", "Old")
	conv1.LastMessageAt = now.Add(-3 * time.Hour)
	conv2 := newTestConv(uid(), userA, uid(), "direct", "Mid")
	conv2.LastMessageAt = now.Add(-1 * time.Hour)
	conv3 := newTestConv(uid(), uid(), userA, "direct", "Recent")
	conv3.LastMessageAt = now

	require.NoError(t, db.Conversations.Create(ctx, conv1))
	require.NoError(t, db.Conversations.Create(ctx, conv2))
	require.NoError(t, db.Conversations.Create(ctx, conv3))

	// Get first page: limit=2.
	results, err := db.Conversations.GetByUser(ctx, userA, 0, 2)
	require.NoError(t, err)
	require.Len(t, results, 2)
	// Should be ordered by LastMessageAt DESC: Recent, Mid.
	assert.Equal(t, conv3.ID, results[0].ID)
	assert.Equal(t, conv2.ID, results[1].ID)

	// Get second page: offset=2, limit=2.
	results, err = db.Conversations.GetByUser(ctx, userA, 2, 2)
	require.NoError(t, err)
	require.Len(t, results, 1)
	assert.Equal(t, conv1.ID, results[0].ID)

	// Soft-delete conv2 and verify it's excluded.
	require.NoError(t, db.Conversations.Delete(ctx, conv2.ID))
	results, err = db.Conversations.GetByUser(ctx, userA, 0, 10)
	require.NoError(t, err)
	require.Len(t, results, 2)
	for _, r := range results {
		assert.NotEqual(t, conv2.ID, r.ID)
	}
}

func TestConversationStore_Update_Success(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "Original")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	conv.Title = "Updated"
	conv.Pinned = true
	require.NoError(t, db.Conversations.Update(ctx, conv))

	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, "Updated", got.Title)
	assert.True(t, got.Pinned)
}

func TestConversationStore_Delete_CascadeSoftDelete(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "ToDelete")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	// Create messages in this conversation.
	msg1 := newTestMsg(uid(), uid(), conv.ID, 1, "sender", "msg1")
	msg2 := newTestMsg(uid(), uid(), conv.ID, 2, "sender", "msg2")
	require.NoError(t, db.Messages.Create(ctx, msg1))
	require.NoError(t, db.Messages.Create(ctx, msg2))

	// Delete the conversation (D-013: cascade soft-delete).
	require.NoError(t, db.Conversations.Delete(ctx, conv.ID))

	// Conversation should not be found via regular Get.
	_, err := db.Conversations.Get(ctx, conv.ID)
	assert.ErrorIs(t, err, ErrNotFound)

	// But should be found via GetUnscoped.
	unscoped, err := db.Conversations.GetUnscoped(ctx, conv.ID)
	require.NoError(t, err)
	assert.NotNil(t, unscoped.DeletedAt)

	// Messages should also be soft-deleted.
	_, err = db.Messages.Get(ctx, msg1.ID)
	assert.ErrorIs(t, err, ErrNotFound)
	_, err = db.Messages.Get(ctx, msg2.ID)
	assert.ErrorIs(t, err, ErrNotFound)
}

func TestConversationStore_Delete_NotFound(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	err := db.Conversations.Delete(ctx, "nonexistent-id")
	assert.ErrorIs(t, err, ErrNotFound)
}

func TestConversationStore_Restore_CascadeRestore(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "ToRestore")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	msg1 := newTestMsg(uid(), uid(), conv.ID, 1, "sender", "msg1")
	msg2 := newTestMsg(uid(), uid(), conv.ID, 2, "sender", "msg2")
	require.NoError(t, db.Messages.Create(ctx, msg1))
	require.NoError(t, db.Messages.Create(ctx, msg2))

	// Delete then restore (D-015: cascade restore).
	require.NoError(t, db.Conversations.Delete(ctx, conv.ID))
	require.NoError(t, db.Conversations.Restore(ctx, conv.ID))

	// Conversation should be back.
	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, conv.ID, got.ID)
	assert.True(t, got.DeletedAt.Time.IsZero())

	// Messages should also be restored.
	gotMsg1, err := db.Messages.Get(ctx, msg1.ID)
	require.NoError(t, err)
	assert.True(t, gotMsg1.DeletedAt.Time.IsZero())

	gotMsg2, err := db.Messages.Get(ctx, msg2.ID)
	require.NoError(t, err)
	assert.True(t, gotMsg2.DeletedAt.Time.IsZero())
}

func TestConversationStore_Restore_NotFound(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	err := db.Conversations.Restore(ctx, "nonexistent-id")
	assert.ErrorIs(t, err, ErrNotFound)
}

func TestConversationStore_UpdateLastRead_MaxSemantics(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	uid1 := uid()
	uid2 := uid()
	conv := newTestConv(uid(), uid1, uid2, "direct", "ReadTest")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	// Set last read to 5 for uid1.
	require.NoError(t, db.Conversations.UpdateLastRead(ctx, conv.ID, uid1, 5))

	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, uint32(5), got.LastReadMessageID1)

	// Try to set to 3 — should stay at 5 (MAX semantics, D-012).
	require.NoError(t, db.Conversations.UpdateLastRead(ctx, conv.ID, uid1, 3))

	got, err = db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, uint32(5), got.LastReadMessageID1)

	// Set to 10 — should advance to 10.
	require.NoError(t, db.Conversations.UpdateLastRead(ctx, conv.ID, uid1, 10))

	got, err = db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, uint32(10), got.LastReadMessageID1)

	// Test uid2 column.
	require.NoError(t, db.Conversations.UpdateLastRead(ctx, conv.ID, uid2, 7))
	got, err = db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, uint32(7), got.LastReadMessageID2)
}

func TestConversationStore_UpdateLastRead_NotMember(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "ReadTest")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	err := db.Conversations.UpdateLastRead(ctx, conv.ID, "non-member-user", 5)
	assert.ErrorIs(t, err, ErrNotFound)
}

func TestConversationStore_SearchByTitle_LikeEscaping(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	user := uid()
	// Create conversations with special LIKE characters in titles.
	conv100pct := newTestConv(uid(), user, uid(), "direct", "100% done")
	conv1001 := newTestConv(uid(), user, uid(), "direct", "1001 items")
	convA_B := newTestConv(uid(), user, uid(), "direct", "a_b test")
	convACB := newTestConv(uid(), user, uid(), "direct", "acb test")
	convPipe := newTestConv(uid(), user, uid(), "direct", "pipe|char")

	require.NoError(t, db.Conversations.Create(ctx, conv100pct))
	require.NoError(t, db.Conversations.Create(ctx, conv1001))
	require.NoError(t, db.Conversations.Create(ctx, convA_B))
	require.NoError(t, db.Conversations.Create(ctx, convACB))
	require.NoError(t, db.Conversations.Create(ctx, convPipe))

	// Search for "100%" — should only match "100% done", not "1001 items".
	results, err := db.Conversations.SearchByTitle(ctx, user, "100%", 10)
	require.NoError(t, err)
	require.Len(t, results, 1)
	assert.Equal(t, conv100pct.ID, results[0].ID)

	// Search for "a_b" — should only match "a_b test", not "acb test".
	results, err = db.Conversations.SearchByTitle(ctx, user, "a_b", 10)
	require.NoError(t, err)
	require.Len(t, results, 1)
	assert.Equal(t, convA_B.ID, results[0].ID)

	// Search for "pipe|char" — should match the pipe conversation.
	results, err = db.Conversations.SearchByTitle(ctx, user, "pipe|char", 10)
	require.NoError(t, err)
	require.Len(t, results, 1)
	assert.Equal(t, convPipe.ID, results[0].ID)

	// Search for empty string returns empty.
	results, err = db.Conversations.SearchByTitle(ctx, user, "", 10)
	require.NoError(t, err)
	assert.Empty(t, results)
}

func TestConversationStore_GetUnscoped_IncludesSoftDeleted(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "SoftDel")
	require.NoError(t, db.Conversations.Create(ctx, conv))
	require.NoError(t, db.Conversations.Delete(ctx, conv.ID))

	// Regular Get should fail.
	_, err := db.Conversations.Get(ctx, conv.ID)
	assert.ErrorIs(t, err, ErrNotFound)

	// GetUnscoped should succeed and include the soft-deleted record.
	got, err := db.Conversations.GetUnscoped(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, conv.ID, got.ID)
	assert.NotNil(t, got.DeletedAt)
	assert.False(t, got.DeletedAt.Time.IsZero())
}

func TestConversationStore_UpdateLastMessage(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "LastMsg")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	newTime := time.Now().Add(1 * time.Hour)
	require.NoError(t, db.Conversations.UpdateLastMessage(ctx, conv.ID, newTime, 42))

	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, uint32(42), got.LastProcessedMessageID)
	// LastMessageAt may have slight precision loss due to SQLite storage.
	assert.WithinDuration(t, newTime, got.LastMessageAt, time.Second)
}

func TestConversationStore_UpdateLastMessage_NotFound(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	err := db.Conversations.UpdateLastMessage(ctx, "nonexistent", time.Now(), 1)
	assert.ErrorIs(t, err, ErrNotFound)
}

// Verify that ConversationStore errors wrap correctly.
func TestConversationStore_ErrorWrapping(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	_, err := db.Conversations.Get(ctx, "nonexistent")
	assert.True(t, errors.Is(err, ErrNotFound))
}

func TestConversationStore_GetByUsers_NotFound(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	_, err := db.Conversations.GetByUsers(ctx, uid(), uid())
	assert.ErrorIs(t, err, ErrNotFound)
}

func TestConversationStore_UpdateLastRead_ConversationNotFound(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	err := db.Conversations.UpdateLastRead(ctx, "nonexistent-conv-id", uid(), 5)
	assert.ErrorIs(t, err, ErrNotFound)
}

func TestConversationStore_Restore_NotDeleted(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "NotDeleted")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	// Restore on a conversation that is NOT deleted is idempotent (D-015):
	// returns nil without error.
	err := db.Conversations.Restore(ctx, conv.ID)
	assert.NoError(t, err)
}

// ---------------------------------------------------------------------------
// Upsert tests (Bug #5 — D-045)
// ---------------------------------------------------------------------------

func TestConversationStore_Upsert_Create(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "UpsertNew")
	// Upsert on a non-existing record should create it.
	err := db.Conversations.Upsert(ctx, conv)
	require.NoError(t, err)

	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, conv.ID, got.ID)
	assert.Equal(t, "UpsertNew", got.Title)
}

func TestConversationStore_Upsert_Update(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "Original")
	require.NoError(t, db.Conversations.Create(ctx, conv))

	// Upsert on an existing record should update it.
	conv.Title = "Updated"
	conv.Pinned = true
	err := db.Conversations.Upsert(ctx, conv)
	require.NoError(t, err)

	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, "Updated", got.Title)
	assert.True(t, got.Pinned)
}

func TestConversationStore_Upsert_Idempotent(t *testing.T) {
	db := setupTestDB(t)
	ctx := context.Background()
	cleanAll(t, db, ctx)

	conv := newTestConv(uid(), uid(), uid(), "direct", "Idempotent")
	// Call Upsert twice — should succeed both times without error.
	require.NoError(t, db.Conversations.Upsert(ctx, conv))
	require.NoError(t, db.Conversations.Upsert(ctx, conv))

	// Only one record should exist.
	got, err := db.Conversations.Get(ctx, conv.ID)
	require.NoError(t, err)
	assert.Equal(t, conv.ID, got.ID)
}

File: demo/vue-pure-admin/packages/xyncra-client-core/src/__tests__/db/conversation-store.test.ts
/**
 * ConversationStore unit tests.
 *
 * Tests cover:
 *   - CRUD operations
 *   - C4: UpdateLastRead MAX semantics
 *   - Soft delete and restore with cascade
 *   - Transaction variants (Tx)
 *   - Edge cases: not found, duplicate key, getByUsers ordering
 */

import { ErrNotFound } from '../../errors';
import {
  createConversation,
  createFreshDatabase,
  createMessage,
  resetIdCounter,
} from '../test-helpers';

describe('ConversationStore', () => {
  let db: ReturnType<typeof createFreshDatabase>;

  beforeEach(async () => {
    resetIdCounter();
    db = createFreshDatabase(`test-conv-${Date.now()}-${Math.random()}`);
    await db.open();
  });

  afterEach(async () => {
    await db.delete();
  });

  // ---------------------------------------------------------------------------
  // CRUD
  // ---------------------------------------------------------------------------

  test('create and get', async () => {
    const conv = createConversation({ user_id1: 'alice', user_id2: 'bob' });
    await db.conversationsStore.create(conv);

    const retrieved = await db.conversationsStore.get(conv.id);
    expect(retrieved).toBeDefined();
    expect(retrieved!.user_id1).toBe('alice');
    expect(retrieved!.user_id2).toBe('bob');
  });

  test('get returns undefined for non-existent id', async () => {
    const result = await db.conversationsStore.get('non-existent');
    expect(result).toBeUndefined();
  });

  test('get excludes soft-deleted records', async () => {
    const conv = createConversation({ deleted_at: new Date() });
    await db.conversations.add(conv);

    const result = await db.conversationsStore.get(conv.id);
    expect(result).toBeUndefined();
  });

  test('getUnscoped includes soft-deleted records', async () => {
    const conv = createConversation({ deleted_at: new Date() });
    await db.conversations.add(conv);

    const result = await db.conversationsStore.getUnscoped(conv.id);
    expect(result).toBeDefined();
    expect(result!.deleted_at).not.toBeNull();
  });

  // ---------------------------------------------------------------------------
  // getByUsers
  // ---------------------------------------------------------------------------

  test('getByUsers finds conversation in either order', async () => {
    const conv = createConversation({ user_id1: 'alice', user_id2: 'bob' });
    await db.conversationsStore.create(conv);

    // Forward order
    let found = await db.conversationsStore.getByUsers('alice', 'bob');
    expect(found).toBeDefined();
    expect(found!.id).toBe(conv.id);

    // Reverse order
    found = await db.conversationsStore.getByUsers('bob', 'alice');
    expect(found).toBeDefined();
    expect(found!.id).toBe(conv.id);
  });

  test('getByUsers returns undefined when not found', async () => {
    const found = await db.conversationsStore.getByUsers('nobody', 'here');
    expect(found).toBeUndefined();
  });

  test('getByUsers excludes soft-deleted', async () => {
    const conv = createConversation({
      user_id1: 'alice',
      user_id2: 'bob',
      deleted_at: new Date(),
    });
    await db.conversations.add(conv);

    const found = await db.conversationsStore.getByUsers('alice', 'bob');
    expect(found).toBeUndefined();
  });

  // ---------------------------------------------------------------------------
  // getByUser
  // ---------------------------------------------------------------------------

  test('getByUser returns conversations sorted by last_message_at desc', async () => {
    const conv1 = createConversation({
      user_id1: 'alice',
      user_id2: 'bob',
      last_message_at: new Date('2026-01-01'),
    });
    const conv2 = createConversation({
      user_id1: 'alice',
      user_id2: 'charlie',
      last_message_at: new Date('2026-06-01'),
    });
    await db.conversationsStore.create(conv1);
    await db.conversationsStore.create(conv2);

    const result = await db.conversationsStore.getByUser('alice', 0, 20);
    expect(result).toHaveLength(2);
    expect(result[0].id).toBe(conv2.id); // newer first
  });

  test('getByUser excludes soft-deleted', async () => {
    const conv1 = createConversation({ user_id1: 'alice', user_id2: 'bob' });
    const conv2 = createConversation({
      user_id1: 'alice',
      user_id2: 'charlie',
      deleted_at: new Date(),
    });
    await db.conversationsStore.create(conv1);
    await db.conversations.add(conv2);

    const result = await db.conversationsStore.getByUser('alice', 0, 20);
    expect(result).toHaveLength(1);
    expect(result[0].id).toBe(conv1.id);
  });

  // ---------------------------------------------------------------------------
  // C4: UpdateLastRead MAX semantics
  // ---------------------------------------------------------------------------

  test('C4: updateLastRead only advances forward for user_id1', async () => {
    const conv = createConversation({
      user_id1: 'alice',
      user_id2: 'bob',
      last_read_message_id1: 0,
    });
    await db.conversationsStore.create(conv);

    // Advance to 10
    await db.conversationsStore.updateLastRead(conv.id, 'alice', 10);
    let updated = await db.conversationsStore.get(conv.id);
    expect(updated!.last_read_message_id1).toBe(10);

    // Advance to 20 (should succeed)
    await db.conversationsStore.updateLastRead(conv.id, 'alice', 20);
    updated = await db.conversationsStore.get(conv.id);
    expect(updated!.last_read_message_id1).toBe(20);

    // Try to go back to 5 (should be rejected - MAX semantics)
    await db.conversationsStore.updateLastRead(conv.id, 'alice', 5);
    updated = await db.conversationsStore.get(conv.id);
    expect(updated!.last_read_message_id1).toBe(20); // unchanged
  });

  test('C4: updateLastRead works for user_id2', async () => {
    const conv = createConversation({
      user_id1: 'alice',
      user_id2: 'bob',
      last_read_message_id2: 0,
    });
    await db.conversationsStore.create(conv);

    await db.conversationsStore.updateLastRead(conv.id, 'bob', 15);
    const updated = await db.conversationsStore.get(conv.id);
    expect(updated!.last_read_message_id2).toBe(15);
  });

  test('updateLastRead throws ErrNotFound for unknown user', async () => {
    const conv = createConversation({ user_id1: 'alice', user_id2: 'bob' });
    await db.conversationsStore.create(conv);

    await expect(
      db.conversationsStore.updateLastRead(conv.id, 'charlie', 10),
    ).rejects.toBe(ErrNotFound);
  });

  // ---------------------------------------------------------------------------
  // Soft delete and restore (cascade)
  // ---------------------------------------------------------------------------

  test('delete cascade soft-deletes conversation and messages', async () => {
    const conv = createConversation();
    await db.conversationsStore.create(conv);

    const msg = createMessage({ conversation_id: conv.id });
    await db.messagesStore.create(msg);

    await db.conversationsStore.delete(conv.id);

    // Conversation is no longer visible
    const convResult = await db.conversationsStore.get(conv.id);
    expect(convResult).toBeUndefined();

    // But visible via getUnscoped
    const convUnscoped = await db.conversationsStore.getUnscoped(conv.id);
    expect(convUnscoped).toBeDefined();
    expect(convUnscoped!.deleted_at).not.toBeNull();

    // Message is also soft-deleted
    const msgResult = await db.messagesStore.get(msg.id);
    expect(msgResult).toBeUndefined();
  });

  test('restore cascade restores conversation and messages', async () => {
    const conv = createConversation();
    await db.conversationsStore.create(conv);

    const msg = createMessage({ conversation_id: conv.id });
    await db.messagesStore.create(msg);

    // Delete first
    await db.conversationsStore.delete(conv.id);

    // Then restore
    await db.conversationsStore.restore(conv.id);

    // Conversation is visible again
    const convResult = await db.conversationsStore.get(conv.id);
    expect(convResult).toBeDefined();
    expect(convResult!.deleted_at).toBeNull();

    // Message is also restored
    const msgResult = await db.messagesStore.get(msg.id);
    expect(msgResult).toBeDefined();
  });

  test('restore is idempotent on non-deleted conversation', async () => {
    const conv = createConversation();
    await db.conversationsStore.create(conv);

    // Should not throw
    await db.conversationsStore.restore(conv.id);

    const result = await db.conversationsStore.get(conv.id);
    expect(result).toBeDefined();
  });

  test('delete throws ErrNotFound for non-existent', async () => {
    await expect(db.conversationsStore.delete('nope')).rejects.toBe(
      ErrNotFound,
    );
  });

  // ---------------------------------------------------------------------------
  // upsert
  // ---------------------------------------------------------------------------

  test('upsert creates new record', async () => {
    const conv = createConversation({ title: 'New' });
    await db.conversationsStore.upsert(conv);

    const result = await db.conversationsStore.get(conv.id);
    expect(result).toBeDefined();
    expect(result!.title).toBe('New');
  });

  test('upsert updates existing record', async () => {
    const conv = createConversation({ title: 'Original' });
    await db.conversationsStore.create(conv);

    conv.title = 'Updated';
    await db.conversationsStore.upsert(conv);

    const result = await db.conversationsStore.get(conv.id);
    expect(result!.title).toBe('Updated');
  });

  // ---------------------------------------------------------------------------
  // searchByTitle
  // ---------------------------------------------------------------------------

  test('searchByTitle finds case-insensitive matches', async () => {
    const conv1 = createConversation({
      user_id1: 'alice',
      user_id2: 'bob',
      title: 'Hello World',
    });
    const conv2 = createConversation({
      user_id1: 'alice',
      user_id2: 'charlie',
      title: 'Goodbye',
    });
    await db.conversationsStore.create(conv1);
    await db.conversationsStore.create(conv2);

    const results = await db.conversationsStore.searchByTitle(
      'alice',
      'hello',
      10,
    );
    expect(results).toHaveLength(1);
    expect(results[0].id).toBe(conv1.id);
  });

  test('searchByTitle returns empty for empty query', async () => {
    const conv = createConversation({ user_id1: 'alice', title: 'Test' });
    await db.conversationsStore.create(conv);

    const results = await db.conversationsStore.searchByTitle('alice', '', 10);
    expect(results).toHaveLength(0);
  });

  // ---------------------------------------------------------------------------
  // updateLastMessage
  // ---------------------------------------------------------------------------

  test('updateLastMessage updates fields', async () => {
    const conv = createConversation();
    await db.conversationsStore.create(conv);

    const newDate = new Date('2026-07-01');
    await db.conversationsStore.updateLastMessage(conv.id, newDate, 42);

    const result = await db.conversationsStore.get(conv.id);
    expect(result!.last_message_at).toEqual(newDate);
    expect(result!.last_processed_message_id).toBe(42);
  });

  test('updateLastMessage throws ErrNotFound for unknown', async () => {
    await expect(
      db.conversationsStore.updateLastMessage('nope', new Date(), 1),
    ).rejects.toBe(ErrNotFound);
  });
});
